packages/core/src/submodules/cbor/cbor-decode.ts

```diff
@@ -0,0 +1,477 @@
+    if (to - at < _offset) {
+      throw new Error(`countLength ${countLength} greater than remaining buf len.`);
+    }
+    const countIndex = at + 1;
+
+    if (countLength === 1) {
+      return payload[countIndex];
+    } else if (countLength === 2) {
+      return dataView.getUint16(countIndex);
+    } else if (countLength === 4) {
+      return dataView.getUint32(countIndex);
+    }
+    return demote(dataView.getBigUint64(countIndex));
+  }
+
+  throw new Error(`unexpected minor value ${minor}.`);
+}
+
+function decodeUtf8String(at: Uint32, to: Uint32): string {
+  const length = decodeCount(at, to);
+  const offset = _offset;
+  at += offset;
+  if (to - at < length) {
+    throw new Error(`string len ${length} greater than remaining buf len.`);
+  }
+  const value = bytesToUtf8(payload, at, at + length);
+  _offset = offset + length;
+  return value;
+}
+
+function decodeUtf8StringIndefinite(at: Uint32, to: Uint32): string {
+  at += 1;
+  const vector = [];
+  for (const base = at; at < to; ) {
+    if (payload[at] === 0b1111_1111) {
+      const data = alloc(vector.length);
+      data.set(vector, 0);
+      _offset = at - base + 2;
+      return bytesToUtf8(data, 0, data.length);
+    }
+    const major = (payload[at] & 0b1110_0000) >> 5;
+    const minor = payload[at] & 0b0001_1111;
+    if (major !== majorUtf8String) {
+      throw new Error(`unexpected major type ${major} in indefinite string.`);
+    }
+    if (minor === minorIndefinite) {
+      throw new Error("nested indefinite string.");
+    }
+    const bytes = decodeUnstructuredByteString(at, to);
+    const length = _offset;
+    at += length;
+    vector.push(...bytes);
+  }
+  throw new Error("expected break marker.");
+}
+
+function decodeUnstructuredByteString(at: Uint32, to: Uint32): CborUnstructuredByteStringType {
+  const length = decodeCount(at, to);
+  const offset = _offset;
+
+  at += offset;
+  if (to - at < length) {
+    throw new Error(`unstructured byte string len ${length} greater than remaining buf len.`);
+  }
+
+  const value = payload.subarray(at, at + length);
+  _offset = offset + length;
+  return value;
+}
+
+function decodeUnstructuredByteStringIndefinite(at: Uint32, to: Uint32): CborUnstructuredByteStringType {
+  at += 1;
+  const vector = [];
+
+  for (const base = at; at < to; ) {
+    if (payload[at] === 0b1111_1111) {
+      const data = alloc(vector.length);
+      data.set(vector, 0);
+      _offset = at - base + 2;
+      return data;
+    }
+
+    const major = (payload[at] & 0b1110_0000) >> 5;
+    const minor = payload[at] & 0b0001_1111;
+    if (major !== majorUnstructuredByteString) {
+      throw new Error(`unexpected major type ${major} in indefinite string.`);
+    }
+    if (minor === minorIndefinite) {
+      throw new Error("nested indefinite string.");
+    }
+
+    const bytes = decodeUnstructuredByteString(at, to);
+    const length = _offset;
+    at += length;
+    vector.push(...bytes);
```

Comment:
Not sure how this works in JS - does spreading `bytes` create a new array?

Reply:
if im not wrong, it spreads the elements of the `bytes` array as individual arguments to `push()`.

Reply:
semantically, it doesn't require the creation of an intermediate temporary array, but I believe the implementation does so.

I'm switching this to an index for loop, since `...args` spread is known to crash earlier than index for-loops.